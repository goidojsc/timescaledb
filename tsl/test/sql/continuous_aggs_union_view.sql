-- This file and its contents are licensed under the Apache License 2.0.
-- Please see the included NOTICE for copyright information and
-- LICENSE-APACHE for a copy of the license.

-- disable background workers to make results reproducible
\c :TEST_DBNAME :ROLE_SUPERUSER
SELECT _timescaledb_internal.stop_background_workers();
\c :TEST_DBNAME :ROLE_DEFAULT_PERM_USER

CREATE TABLE metrics(f1 int, f2 int, time timestamptz NOT NULL, device_id int, value float);
SELECT create_hypertable('metrics','time');
ALTER TABLE metrics DROP COLUMN f1;

INSERT INTO metrics(time, device_id, value) SELECT '2000-01-01'::timestamptz, 1, 0.5;
INSERT INTO metrics(time, device_id, value) SELECT '2000-01-01'::timestamptz, 2, 0.5;
INSERT INTO metrics(time, device_id, value) SELECT '2000-01-01'::timestamptz, 3, 0.5;

--
-- test switching continuous agg view between different modes
--

-- check default view for new continuous aggregate
CREATE VIEW metrics_summary WITH (timescaledb.continuous) AS SELECT time_bucket('1d',time), avg(value) FROM metrics GROUP BY 1;
ALTER TABLE metrics DROP COLUMN f2;

-- this should be union view
SELECT pg_get_viewdef('metrics_summary',true);
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;

-- downgrade view to non-union view
ALTER VIEW metrics_summary SET (timescaledb.materialized_only=true);
-- this should be view without union
SELECT pg_get_viewdef('metrics_summary',true);
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;

-- upgrade view to union view again
ALTER VIEW metrics_summary SET (timescaledb.materialized_only=false);
-- this should be union view
SELECT pg_get_viewdef('metrics_summary',true);
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;

-- try upgrade view to union view that is already union view
ALTER VIEW metrics_summary SET (timescaledb.materialized_only=false);
-- this should be union view
SELECT pg_get_viewdef('metrics_summary',true);
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;

-- refresh
REFRESH MATERIALIZED VIEW metrics_summary;
-- result should not change after refresh for union view
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;

-- downgrade view to non-union view
ALTER VIEW metrics_summary SET (timescaledb.materialized_only=true);
-- this should be view without union
SELECT pg_get_viewdef('metrics_summary',true);
-- view should have results now after refresh
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;

DROP VIEW metrics_summary CASCADE;

-- check default view for new continuous aggregate with materialized_only to true
CREATE VIEW metrics_summary WITH (timescaledb.continuous, timescaledb.materialized_only=true) AS SELECT time_bucket('1d',time), avg(value) FROM metrics GROUP BY 1;

-- this should be view without union
SELECT pg_get_viewdef('metrics_summary',true);
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;

-- upgrade view to union view
ALTER VIEW metrics_summary SET (timescaledb.materialized_only=false);
-- this should be union view
SELECT pg_get_viewdef('metrics_summary',true);
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;

-- downgrade view to non-union view
ALTER VIEW metrics_summary SET (timescaledb.materialized_only=true);
-- this should be view without union
SELECT pg_get_viewdef('metrics_summary',true);
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;

DROP VIEW metrics_summary CASCADE;

--
-- test queries on union view
--

CREATE VIEW metrics_summary WITH (timescaledb.continuous, timescaledb.materialized_only=true) AS SELECT time_bucket('1d',time), avg(value) FROM metrics GROUP BY 1;

-- query should not have results since cagg is materialized only and no refresh has happened yet
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;

ALTER VIEW metrics_summary SET (timescaledb.materialized_only=false);

-- after switch to union view all results should be returned
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;

REFRESH MATERIALIZED VIEW metrics_summary;
ALTER VIEW metrics_summary SET (timescaledb.materialized_only=true);

-- materialized only view should return data now too because refresh has happened
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;

-- add some more data
INSERT INTO metrics(time, device_id, value) SELECT '2000-02-01'::timestamptz, 1, 0.1;
INSERT INTO metrics(time, device_id, value) SELECT '2000-02-01'::timestamptz, 2, 0.2;
INSERT INTO metrics(time, device_id, value) SELECT '2000-02-01'::timestamptz, 3, 0.3;

-- materialized only view should not have new data yet
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;

-- but union view should
ALTER VIEW metrics_summary SET (timescaledb.materialized_only=false);
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;

-- and after refresh non union view should have new data too
REFRESH MATERIALIZED VIEW metrics_summary;
ALTER VIEW metrics_summary SET (timescaledb.materialized_only=true);
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;

