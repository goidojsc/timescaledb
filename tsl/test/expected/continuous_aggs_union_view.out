-- This file and its contents are licensed under the Apache License 2.0.
-- Please see the included NOTICE for copyright information and
-- LICENSE-APACHE for a copy of the license.
-- disable background workers to make results reproducible
\c :TEST_DBNAME :ROLE_SUPERUSER
SELECT _timescaledb_internal.stop_background_workers();
 stop_background_workers 
-------------------------
 t
(1 row)

\c :TEST_DBNAME :ROLE_DEFAULT_PERM_USER
CREATE TABLE metrics(f1 int, f2 int, time timestamptz NOT NULL, device_id int, value float);
SELECT create_hypertable('metrics','time');
  create_hypertable   
----------------------
 (1,public,metrics,t)
(1 row)

ALTER TABLE metrics DROP COLUMN f1;
INSERT INTO metrics(time, device_id, value) SELECT '2000-01-01'::timestamptz, 1, 0.5;
INSERT INTO metrics(time, device_id, value) SELECT '2000-01-01'::timestamptz, 2, 0.5;
INSERT INTO metrics(time, device_id, value) SELECT '2000-01-01'::timestamptz, 3, 0.5;
--
-- test switching continuous agg view between different modes
--
-- check default view for new continuous aggregate
CREATE VIEW metrics_summary WITH (timescaledb.continuous) AS SELECT time_bucket('1d',time), avg(value) FROM metrics GROUP BY 1;
ALTER TABLE metrics DROP COLUMN f2;
-- this should be union view
SELECT pg_get_viewdef('metrics_summary',true);
                                                                                          pg_get_viewdef                                                                                           
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  SELECT _materialized_hypertable_2.time_bucket,                                                                                                                                                  +
     _timescaledb_internal.finalize_agg('avg(double precision)'::text, NULL::name, NULL::name, '{{pg_catalog,float8}}'::name[], _materialized_hypertable_2.agg_2_2, NULL::double precision) AS avg+
    FROM _timescaledb_internal._materialized_hypertable_2                                                                                                                                         +
   WHERE _materialized_hypertable_2.time_bucket < COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(1)), '-infinity'::timestamp with time zone)                    +
   GROUP BY _materialized_hypertable_2.time_bucket                                                                                                                                                +
 UNION ALL                                                                                                                                                                                        +
  SELECT time_bucket('@ 1 day'::interval, metrics."time") AS time_bucket,                                                                                                                         +
     avg(metrics.value) AS avg                                                                                                                                                                    +
    FROM metrics                                                                                                                                                                                  +
   WHERE metrics."time" > COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(1)), '-infinity'::timestamp with time zone)                                            +
   GROUP BY (time_bucket('@ 1 day'::interval, metrics."time"));
(1 row)

SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;
         time_bucket          | avg 
------------------------------+-----
 Fri Dec 31 16:00:00 1999 PST | 0.5
(1 row)

-- downgrade view to non-union view
ALTER VIEW metrics_summary SET (timescaledb.materialized_only=true);
-- this should be view without union
SELECT pg_get_viewdef('metrics_summary',true);
                                                                                          pg_get_viewdef                                                                                           
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  SELECT _materialized_hypertable_2.time_bucket,                                                                                                                                                  +
     _timescaledb_internal.finalize_agg('avg(double precision)'::text, NULL::name, NULL::name, '{{pg_catalog,float8}}'::name[], _materialized_hypertable_2.agg_2_2, NULL::double precision) AS avg+
    FROM _timescaledb_internal._materialized_hypertable_2                                                                                                                                         +
   GROUP BY _materialized_hypertable_2.time_bucket;
(1 row)

SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;
 time_bucket | avg 
-------------+-----
(0 rows)

-- upgrade view to union view again
ALTER VIEW metrics_summary SET (timescaledb.materialized_only=false);
-- this should be union view
SELECT pg_get_viewdef('metrics_summary',true);
                                                                                          pg_get_viewdef                                                                                           
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  SELECT _materialized_hypertable_2.time_bucket,                                                                                                                                                  +
     _timescaledb_internal.finalize_agg('avg(double precision)'::text, NULL::name, NULL::name, '{{pg_catalog,float8}}'::name[], _materialized_hypertable_2.agg_2_2, NULL::double precision) AS avg+
    FROM _timescaledb_internal._materialized_hypertable_2                                                                                                                                         +
   WHERE _materialized_hypertable_2.time_bucket < COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(1)), '-infinity'::timestamp with time zone)                    +
   GROUP BY _materialized_hypertable_2.time_bucket                                                                                                                                                +
 UNION ALL                                                                                                                                                                                        +
  SELECT time_bucket('@ 1 day'::interval, metrics."time") AS time_bucket,                                                                                                                         +
     avg(metrics.value) AS avg                                                                                                                                                                    +
    FROM metrics                                                                                                                                                                                  +
   WHERE metrics."time" > COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(1)), '-infinity'::timestamp with time zone)                                            +
   GROUP BY (time_bucket('@ 1 day'::interval, metrics."time"));
(1 row)

SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;
         time_bucket          | avg 
------------------------------+-----
 Fri Dec 31 16:00:00 1999 PST | 0.5
(1 row)

-- try upgrade view to union view that is already union view
ALTER VIEW metrics_summary SET (timescaledb.materialized_only=false);
-- this should be union view
SELECT pg_get_viewdef('metrics_summary',true);
                                                                                          pg_get_viewdef                                                                                           
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  SELECT _materialized_hypertable_2.time_bucket,                                                                                                                                                  +
     _timescaledb_internal.finalize_agg('avg(double precision)'::text, NULL::name, NULL::name, '{{pg_catalog,float8}}'::name[], _materialized_hypertable_2.agg_2_2, NULL::double precision) AS avg+
    FROM _timescaledb_internal._materialized_hypertable_2                                                                                                                                         +
   WHERE _materialized_hypertable_2.time_bucket < COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(1)), '-infinity'::timestamp with time zone)                    +
   GROUP BY _materialized_hypertable_2.time_bucket                                                                                                                                                +
 UNION ALL                                                                                                                                                                                        +
  SELECT time_bucket('@ 1 day'::interval, metrics."time") AS time_bucket,                                                                                                                         +
     avg(metrics.value) AS avg                                                                                                                                                                    +
    FROM metrics                                                                                                                                                                                  +
   WHERE metrics."time" > COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(1)), '-infinity'::timestamp with time zone)                                            +
   GROUP BY (time_bucket('@ 1 day'::interval, metrics."time"));
(1 row)

SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;
         time_bucket          | avg 
------------------------------+-----
 Fri Dec 31 16:00:00 1999 PST | 0.5
(1 row)

-- refresh
REFRESH MATERIALIZED VIEW metrics_summary;
INFO:  new materialization range for public.metrics (time column time) larger than allowed in one run, truncating Sat Feb 22 16:00:00 2020 PST to Thu Jan 20 16:00:00 2000 PST
INFO:  materializing continuous aggregate public.metrics_summary: nothing to invalidate, new range up to Thu Jan 20 16:00:00 2000 PST
WARNING:  REFRESH did not materialize the entire range since it was limited by the max_interval_per_job setting
-- result should not change after refresh for union view
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;
         time_bucket          | avg 
------------------------------+-----
 Fri Dec 31 16:00:00 1999 PST | 0.5
(1 row)

-- downgrade view to non-union view
ALTER VIEW metrics_summary SET (timescaledb.materialized_only=true);
-- this should be view without union
SELECT pg_get_viewdef('metrics_summary',true);
                                                                                          pg_get_viewdef                                                                                           
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  SELECT _materialized_hypertable_2.time_bucket,                                                                                                                                                  +
     _timescaledb_internal.finalize_agg('avg(double precision)'::text, NULL::name, NULL::name, '{{pg_catalog,float8}}'::name[], _materialized_hypertable_2.agg_2_2, NULL::double precision) AS avg+
    FROM _timescaledb_internal._materialized_hypertable_2                                                                                                                                         +
   GROUP BY _materialized_hypertable_2.time_bucket;
(1 row)

-- view should have results now after refresh
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;
         time_bucket          | avg 
------------------------------+-----
 Fri Dec 31 16:00:00 1999 PST | 0.5
(1 row)

DROP VIEW metrics_summary CASCADE;
NOTICE:  drop cascades to table _timescaledb_internal._hyper_2_2_chunk
-- check default view for new continuous aggregate with materialized_only to true
CREATE VIEW metrics_summary WITH (timescaledb.continuous, timescaledb.materialized_only=true) AS SELECT time_bucket('1d',time), avg(value) FROM metrics GROUP BY 1;
-- this should be view without union
SELECT pg_get_viewdef('metrics_summary',true);
                                                                                          pg_get_viewdef                                                                                           
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  SELECT _materialized_hypertable_3.time_bucket,                                                                                                                                                  +
     _timescaledb_internal.finalize_agg('avg(double precision)'::text, NULL::name, NULL::name, '{{pg_catalog,float8}}'::name[], _materialized_hypertable_3.agg_2_2, NULL::double precision) AS avg+
    FROM _timescaledb_internal._materialized_hypertable_3                                                                                                                                         +
   GROUP BY _materialized_hypertable_3.time_bucket;
(1 row)

SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;
 time_bucket | avg 
-------------+-----
(0 rows)

-- upgrade view to union view
ALTER VIEW metrics_summary SET (timescaledb.materialized_only=false);
-- this should be union view
SELECT pg_get_viewdef('metrics_summary',true);
                                                                                          pg_get_viewdef                                                                                           
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  SELECT _materialized_hypertable_3.time_bucket,                                                                                                                                                  +
     _timescaledb_internal.finalize_agg('avg(double precision)'::text, NULL::name, NULL::name, '{{pg_catalog,float8}}'::name[], _materialized_hypertable_3.agg_2_2, NULL::double precision) AS avg+
    FROM _timescaledb_internal._materialized_hypertable_3                                                                                                                                         +
   WHERE _materialized_hypertable_3.time_bucket < COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(1)), '-infinity'::timestamp with time zone)                    +
   GROUP BY _materialized_hypertable_3.time_bucket                                                                                                                                                +
 UNION ALL                                                                                                                                                                                        +
  SELECT time_bucket('@ 1 day'::interval, metrics."time") AS time_bucket,                                                                                                                         +
     avg(metrics.value) AS avg                                                                                                                                                                    +
    FROM metrics                                                                                                                                                                                  +
   WHERE metrics."time" > COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(1)), '-infinity'::timestamp with time zone)                                            +
   GROUP BY (time_bucket('@ 1 day'::interval, metrics."time"));
(1 row)

SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;
         time_bucket          | avg 
------------------------------+-----
 Fri Dec 31 16:00:00 1999 PST | 0.5
(1 row)

-- downgrade view to non-union view
ALTER VIEW metrics_summary SET (timescaledb.materialized_only=true);
-- this should be view without union
SELECT pg_get_viewdef('metrics_summary',true);
                                                                                          pg_get_viewdef                                                                                           
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  SELECT _materialized_hypertable_3.time_bucket,                                                                                                                                                  +
     _timescaledb_internal.finalize_agg('avg(double precision)'::text, NULL::name, NULL::name, '{{pg_catalog,float8}}'::name[], _materialized_hypertable_3.agg_2_2, NULL::double precision) AS avg+
    FROM _timescaledb_internal._materialized_hypertable_3                                                                                                                                         +
   GROUP BY _materialized_hypertable_3.time_bucket;
(1 row)

SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;
 time_bucket | avg 
-------------+-----
(0 rows)

DROP VIEW metrics_summary CASCADE;
--
-- test queries on union view
--
CREATE VIEW metrics_summary WITH (timescaledb.continuous, timescaledb.materialized_only=true) AS SELECT time_bucket('1d',time), avg(value) FROM metrics GROUP BY 1;
-- query should not have results since cagg is materialized only and no refresh has happened yet
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;
 time_bucket | avg 
-------------+-----
(0 rows)

ALTER VIEW metrics_summary SET (timescaledb.materialized_only=false);
-- after switch to union view all results should be returned
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;
         time_bucket          | avg 
------------------------------+-----
 Fri Dec 31 16:00:00 1999 PST | 0.5
(1 row)

REFRESH MATERIALIZED VIEW metrics_summary;
INFO:  new materialization range for public.metrics (time column time) larger than allowed in one run, truncating Sat Feb 22 16:00:00 2020 PST to Thu Jan 20 16:00:00 2000 PST
INFO:  materializing continuous aggregate public.metrics_summary: nothing to invalidate, new range up to Thu Jan 20 16:00:00 2000 PST
WARNING:  REFRESH did not materialize the entire range since it was limited by the max_interval_per_job setting
ALTER VIEW metrics_summary SET (timescaledb.materialized_only=true);
-- materialized only view should return data now too because refresh has happened
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;
         time_bucket          | avg 
------------------------------+-----
 Fri Dec 31 16:00:00 1999 PST | 0.5
(1 row)

-- add some more data
INSERT INTO metrics(time, device_id, value) SELECT '2000-02-01'::timestamptz, 1, 0.1;
INSERT INTO metrics(time, device_id, value) SELECT '2000-02-01'::timestamptz, 2, 0.2;
INSERT INTO metrics(time, device_id, value) SELECT '2000-02-01'::timestamptz, 3, 0.3;
-- materialized only view should not have new data yet
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;
         time_bucket          | avg 
------------------------------+-----
 Fri Dec 31 16:00:00 1999 PST | 0.5
(1 row)

-- but union view should
ALTER VIEW metrics_summary SET (timescaledb.materialized_only=false);
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;
         time_bucket          | avg 
------------------------------+-----
 Fri Dec 31 16:00:00 1999 PST | 0.5
 Mon Jan 31 16:00:00 2000 PST | 0.2
(2 rows)

-- and after refresh non union view should have new data too
REFRESH MATERIALIZED VIEW metrics_summary;
INFO:  new materialization range for public.metrics (time column time) larger than allowed in one run, truncating Sat Feb 22 16:00:00 2020 PST to Wed Feb 09 16:00:00 2000 PST
INFO:  materializing continuous aggregate public.metrics_summary: nothing to invalidate, new range up to Wed Feb 09 16:00:00 2000 PST
WARNING:  REFRESH did not materialize the entire range since it was limited by the max_interval_per_job setting
ALTER VIEW metrics_summary SET (timescaledb.materialized_only=true);
SELECT time_bucket,avg FROM metrics_summary ORDER BY 1;
         time_bucket          | avg 
------------------------------+-----
 Fri Dec 31 16:00:00 1999 PST | 0.5
 Mon Jan 31 16:00:00 2000 PST | 0.2
(2 rows)

